Sources/View/CollectionReusableViewSizer.swift

```diff
@@ -0,0 +1,141 @@
+                return .fittingSizeLevel
+            case .fixed:
+                return .required
+            case .custom(_, let priority):
+                return priority
+            }
+        }
+    }
+
+    private lazy var view = View()
+
+    func sizeFor(
+        viewModel: ViewModel,
+        width: Dimension = .compressed,
+        height: Dimension = .compressed,
+        attributes: UICollectionViewLayoutAttributes? = nil
+    ) -> CGSize {
+
+        assert(view.viewModel == nil, "🧟‍♂️ Zombie view model detected! Kill it with fire!")
+
+        view.viewModel = viewModel // the sizer populates the dummy view with the view model
+
+        attributes.then(f: { view.apply($0) })
+
+        let layoutView: UIView
+        switch view {
+        case let cell as UICollectionViewCell:
+            layoutView = cell.contentView // because UIKit reasons
+        default:
+            layoutView = view
+        }
+
+        let size = layoutView.systemLayoutSizeFitting(
+            CGSize(width: width.size, height: height.size),
+            withHorizontalFittingPriority: width.priority,
+            verticalFittingPriority: height.priority
+        )
+
+        view.prepareForReuse() // cleans up the dummy view and sets its viewModel to nil
+
+        return size
+    }
+}
+
+struct ConstantSizerCacheKey: Hashable {}
+
+protocol SizerViewModelView: ReusableViewModelView {
+
+    associatedtype SizerCacheKey: Hashable = ConstantSizerCacheKey
+
+    static func sizerCacheKeyFor(viewModel: ViewModel) -> SizerCacheKey
+}
+
+extension SizerViewModelView where SizerCacheKey == ConstantSizerCacheKey {
+
+    static func sizerCacheKeyFor(viewModel: ViewModel) -> SizerCacheKey {
+
+        return SizerCacheKey()
+    }
+}
+
+final class CollectionReusableViewSizerCache<View: UICollectionReusableView & SizerViewModelView> {
+
+    typealias Sizer = CollectionReusableViewSizer<View>
+
+    private struct Key: Hashable {
+        let key: View.SizerCacheKey
+        let width: Sizer.Dimension
+        let height: Sizer.Dimension
+    }
+
+    private var sizer = Sizer()
+    private var cache = [Key: CGSize]()
+
+    func sizeFor(
+        viewModel: Sizer.ViewModel,
+        width: Sizer.Dimension = .compressed,
+        height: Sizer.Dimension = .compressed,
+        attributes: UICollectionViewLayoutAttributes? = nil
+    ) -> CGSize {
+
+        let key = Key(key: View.sizerCacheKeyFor(viewModel: viewModel), width: width, height: height)
+        if let cachedSize = cache[key] { return cachedSize }
+        let size = sizer.sizeFor(viewModel: viewModel, width: width, height: height, attributes: attributes)
+        cache[key] = size
+        return size
+    }
+}
+
+final class CollectionReusableViewSizerCacheGroup {
+
+    private var sizers = [String: AnyObject]()
+
+    func sizeFor<View: UICollectionReusableView & SizerViewModelView>(
+        _ type: View.Type,
+        viewModel: CollectionReusableViewSizer<View>.ViewModel,
```

Comment:
perhaps we can define a `typealias Sizer = CollectionReusableViewSizer<View>` like above to reduce some of this verbosity 💅

Reply:
Above the generic was on the type while here it's on the func so the typealias would also need to be generic:

```
    typealias Sizer<View: UICollectionReusableView & SizerViewModelView> = CollectionReusableViewSizer<View>

    func sizeFor<View: UICollectionReusableView & SizerViewModelView>(
        _ type: View.Type,
        viewModel: Sizer<View>.ViewModel,
        width: Sizer<View>.Dimension = .compressed,
        height: Sizer<View>.Dimension = .compressed,
        attributes: UICollectionViewLayoutAttributes? = nil
    ) -> CGSize {
```

Not sure we gain much with this, do you think it's still worth it?

Reply:
Oh, I see. Missed that 🙈 never mind then!